op-challenger/runner/prestates.go

```diff
@@ -0,0 +1,114 @@
+	}
+	prestateHash, err := gameImpl.GetAbsolutePrestateHash(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to get absolute prestate hash for %v: %w", gameImplAddr, err)
+	}
+	logger.Info("Using on-chain version of prestate", "prestate", prestateHash)
+	hashFetcher := &HashPrestateFetcher{prestateHash: prestateHash}
+	return hashFetcher.getPrestate(ctx, logger, prestateBaseUrl, prestatePath, dataDir, stateConverter)
+}
+
+type HashPrestateFetcher struct {
+	prestateHash common.Hash
+}
+
+func (f *HashPrestateFetcher) getPrestate(ctx context.Context, _ log.Logger, prestateBaseUrl *url.URL, prestatePath string, dataDir string, stateConverter vm.StateConverter) (string, error) {
+	prestateSource := prestates.NewPrestateSource(
+		prestateBaseUrl,
+		prestatePath,
+		filepath.Join(dataDir, "prestates"),
+		stateConverter)
+
+	prestate, err := prestateSource.PrestatePath(ctx, f.prestateHash)
+	if err != nil {
+		return "", fmt.Errorf("failed to get prestate %v: %w", f.prestateHash, err)
+	}
+	return prestate, nil
+}
+
+// NamedPrestateFetcher downloads a file with a specified name from the prestate base URL and uses it as the prestate.
+// The file is re-downloaded on each run rather than being cached. This makes it possible to run the latest builds
+// from develop.
+type NamedPrestateFetcher struct {
+	filename string
+}
+
+func (f *NamedPrestateFetcher) getPrestate(ctx context.Context, logger log.Logger, prestateBaseUrl *url.URL, _ string, dataDir string, stateConverter vm.StateConverter) (string, error) {
+	targetDir := filepath.Join(dataDir, "prestates")
+	if err := os.MkdirAll(targetDir, 0755); err != nil {
+		return "", fmt.Errorf("error creating prestate dir: %w", err)
+	}
+	prestateUrl := prestateBaseUrl.JoinPath(f.filename)
+	ctx, cancel := context.WithTimeout(ctx, time.Minute)
+	defer cancel()
+
+	req, err := http.NewRequestWithContext(ctx, "GET", prestateUrl.String(), nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to create prestate request: %w", err)
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("failed to fetch prestate from %v: %w", prestateUrl, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("%w from url %v: status %v", prestates.ErrPrestateUnavailable, prestateUrl, resp.StatusCode)
+	}
+
+	targetFile := filepath.Join(targetDir, f.filename)
+	out, err := os.Create(targetFile)
+	if err != nil {
+		return "", fmt.Errorf("failed to create prestate file %v: %w", targetFile, err)
+	}
+	defer out.Close()
+	if _, err := io.Copy(out, resp.Body); err != nil {
+		return "", fmt.Errorf("failed to write prestate to %v: %w", targetFile, err)
+	}
+	proof, _, _, err := stateConverter.ConvertStateToProof(ctx, targetFile)
+	if err != nil {
+		return "", fmt.Errorf("invalid prestate file %v: %w", f.filename, err)
+	}
+	logger.Info("Downloaded named prestate", "filename", f.filename, "prestate", proof.ClaimValue)
+	return targetFile, nil
```

Comment:
Just comparing this to the [MultiPrestateProvider](https://github.com/ethereum-optimism/optimism/blob/1eb6d937f06458a9250573e5c36351679c99a28c/op-challenger/game/fault/trace/prestates/multi.go#L79-L104).  Looks like that method writes a temporary file, deletes it if there is an error, and renames it at the end when it is verified.  Should we do the same thing here?

Or are we streamlining it since the file is always redownloaded?

Reply:
Yeah I figured I'd keep it simple since we will always redownload, so worst case we'll get a failure and leave an invalid file around but then next run we'll overwrite it again anyway.  Whereas the MultiPrestateProvider will assume that if the file is present it is valid so needs to be very sure that it's valid before putting it into its final place.